### DISCUSSION TITLE
Spellchecking based on Tree-Sitter queries?

### DISCUSSION THREAD

**Discussion by connorlay** [helix-editor/helix#3637] [Ideas]
For a bit of background: I've been writing a lot of documentation lately for Phoenix Live View and Helix is my daily driver 🧬 ⚗️ 

Would the maintainers be interested in accepting a PR to Helix core that adds a built-in spellchecker that uses tree-sitter queries to determine what parts of the document to spellcheck? For example, consider the following Elixir code:

```elixir
@doc "This documentation contains a tyyyypo"
def this_function_also_contains_a_tyyyypo do
  # mooooore typooooos
end
```

As a user, I would like the text inside of the comment string and the `@doc` to be spellchecked against my preferred language (in this case, en-US), but not the Elixir identifiers themselves. The tree-sitter queries could look something like this:

```scheme
(comment) @spellcheck
```

This would allow for syntax-aware spell checking for each language and would be customizable for each user. Typos could be represented as diagnostics, possibly even with a list of close matches to a known dictionary?

See https://github.com/lewis6991/spellsitter.nvim for an example of what I am thinking of 🙂

**Comment by David-Else** [comment]
Sounds great, my only thought is that https://github.com/valentjn/ltex-ls has proven to be the ultimate spell and grammar checker, could you integrate that somehow? At the moment it needs some special features to work that are not part of the lsp spec. There is a plugin version for nvim, https://github.com/vigoux/ltex-ls.nvim that implements those features, could it be ported to Helix?

**Reply by connorlay** [reply]
Interesting, I haven't seen the LTeX language server before. I don't see why this could be added to Helix once https://github.com/helix-editor/helix/pull/2507 is merged? Supporting non-LSP features would probably need to be handled by a plugin, or a wrapper similar to how https://github.com/typescript-language-server/typescript-language-server wraps the non-LSP features of `tsserver`.

That being said, I would not personally use such a language server and would rather have something simple and built-in to Helix. I don't want or need an automatic proof reader, all I want is a simple spell checker to help me since I am not the greatest touch typist (or speller) 😅

**Reply by mandarvaze** [reply]
💯 for `ltex-ls` I use it everywhere (Emacs, neovim) I would love to see helix support it.

**Reply by getreu** [reply]
A short blog post about Helix and ltex-ls: [Jens Getreu's blog - Note talking with Helix, Tp-Note and LanguageTool](https://blog.getreu.net/20220828-tp-note-new8/)

Comment in [Discussion #3756 · helix-editor/helix](https://github.com/helix-editor/helix/discussions/3756)

**Comment by the-mikedavis** [comment]
Using tree-sitter to determine what should be spell checked sounds like a good idea to me 👍

Spell-checking in general though is a question mark: there isn't anything built-in at the moment or a way to plug in a checker like aspell. It would probably be worth some discussion whether that even belongs in the core or as a plugin.

I think it would be handy if LSP could be used to provide spell-checking capabilities. Maybe languages could inject a language like we do for tree-sitter-comment or markdown in docs and the spelling language server could act on the injected ranges (although currently injected content can't be adressed by a language server afaik).

**Reply by connorlay** [reply]
@the-mikedavis One approach we could take if we wanted this to be built-in is to use [Enchant](https://abiword.github.io/enchant/) via the [enchant crate](https://crates.io/crates/enchant). Apparently enchant is the spell checker that Emacs uses, and it supports multiple different back-ends so it should be fairly portable and customizable if needed. Should this be in core? Based on the design goal of `"Modern and sane builtin features"` mentioned on https://helix-editor.com I think so 🙂

**Reply by connorlay** [reply]
If we wanted something simpler, there is the [ispell crate](https://docs.rs/ispell/0.3.1/ispell/index.html) that we could use to call `ispell`, `aspell`, or `hunspell`.

**Reply by the-mikedavis** [reply]
I worry that building in spell checking would take a lot of work/code for something that can already be done by LSP (although it would need #2507). Enchant doesn't look like too much code but I'm not sure how the dictionaries fit in - would we need to distribute dictionaries for languages?

With supporting multiple languages, I think spell checking has the possibilty to be too large in scope to be good for core. Like a magit mode or #1976, it would be great if it were in core but scope-wise it may be large enough to be a project on its own.

**Reply by connorlay** [reply]
@the-mikedavis I started a proof-of-concept draft PR exploring what a built-in spell checker could look like using the `ispell` crate and `aspell`: https://github.com/helix-editor/helix/pull/3830 There's still much to do, but you can get the idea of how things look if you want 🙂 

Regardless if you decide this approach makes sense or not to include in core, I'm learning a lot by implementing it 😁

**Comment by kirawi** [comment]
Microsoft recently released a blog post on how they introduced spell checking to Visual Studio: https://devblogs.microsoft.com/visualstudio/visual-studio-spell-checker-preview-now-available/

**Comment by the-mikedavis** [comment]
I've come around to the other side on this since my [last comment :point_up:](https://github.com/helix-editor/helix/discussions/3637#discussioncomment-3531710): I think it's reasonable to have spell-checking as a core feature. Finding system-installed dictionaries is fairly straight-forward and fits well into `helix-loader`, plus it's easy to manually install dictionaries from a source like [JetBrains/hunspell-dictionaries](https://github.com/JetBrains/hunspell-dictionaries) if necessary. Diagnostics for incorrect words can fit squarely into the LSP diagnostics code and we can re-use much of the code-actions code for suggesting corrections and providing a way to allow/forbid words in your personal dictionary. Delegating to a language server works well for prose languages but tokenization is difficult for code (i.e. splitting code into words that can be looked up in the dictionary), and we should be in a nice spot to do this robustly per-language with tree-sitter.

Spell checking also requires some other advancements in core that we want for other features like running tree-sitter queries across language injections (like for [rainbow brackets, textobjects](https://github.com/helix-editor/helix/pull/2857#issuecomment-1694728846), [indents](https://github.com/helix-editor/helix/issues/7364#issuecomment-1598830771)) and segmenting a whole buffer into words efficiently (useful for- but maybe different than [lexical autocomplete](https://github.com/helix-editor/helix/issues/1063)).

Specifically about spell-checking, we really only need the ability to: create an in-memory dictionary from Hunspell `.dic` and `.aff` files, check if a word is in the dictionary, and suggest corrections. Personal dictionaries would be nice to include but could be a follow-up.

The existing options we have (hunspell, nuspell, enchant) are a bit annoying from a build perspective: we would need to depend on the presence of libenchant or hunspell as a library at build-time and use FFI, so spell checking as a feature would probably need to be optional. I'd really like to avoid complicating the build or increasing our dependency on C(++) codebases so as [archseer mentions](https://github.com/helix-editor/helix/pull/6343#issuecomment-1696883645), I'm working on a pure-Rust spell-checker compatible with Hunspell dictionaries. I'll make the repo public and push up an integration branch once the project matures some more.

**Comment by jerabaul29** [comment]
As #2507 has been released as part of 23.10, could / should this discussion be revisited? :) In my experience, good spellcheck also in a coding environment (not just pure markdown or text) per se is already quite useful at catching a number of "stupid" bugs :) .

My questions:

- I guess that there is no reason any longer given #2507  to not use (if need be custom per language) LSP(s) for providing spellcheck on all all kinds of relevant files?
- have some users special experience with using LSP-based spellcheck across file types and have recommendations? Ideally, could a tiny bit of extra config be added to get spellcheck-through-LSP out of the box / could such an example be added to the documentation for n00bs like me to just copy paste? :)
- I understand that @the-mikedavis solution above sounds more attractive on the long run, but would a slightly "ad hoc" solution based on other existing LSPs or similar be possible in the meantime?
- should spellcheck status / support be reported in ```helix --health````? (I guess it does not make sense to report it on a per language basis if it can be shared across languages?)?

**Reply by kirawi** [reply]
I believe the library he's written is already able to do spellchecking. Regarding your question about LSP, yes it is possible. I know that at least one person has written an LSP that does spellchecking in a few hundred lines of code, but for the life of me I cannot find the link to their gist. `ltex-ls` does already offer this, but it is bloated if you only want to use spellchecking.

Edit: All of a sudden, I have found it: https://gist.github.com/sigmaSd/8147a22054675104c70044964152381b

**Reply by jerabaul29** [reply]
That's a really neat little spellcheck LSP in a single file, but if I understand correctly it delegates the spellcheck to aspell, so it is just a wrapper to expose aspell as a LSP server, right? :) Not sure how this make it possible to configure aspell then.

But I could think about giving it a try; do you know if there is a way that I can I add it as a LSP used for all my buffer kinds (possibly giving a list of excluded extensions later on if needed), rather than needing to add by hand an entry for each file extension kind that should use it? :)

I cannot see the library you are mentioning, do you have a link? I was looking on cargo, I can find a wip of pure rust spellcheck https://github.com/pluots/zspell , but it is likely not the one you refer to? :)

**Reply by jerabaul29** [reply]
Sorry, guess it is this one? https://crates.io/crates/spellbook

**Reply by jerabaul29** [reply]
(the github repo linked to there is a 404, so guess this is what is meant by not released publicly yet :) but interesting the crate is already visible :) ).

**Reply by acaloiaro** [reply]
For what it's worth, I use `ltex-ls` for spell checking in `md` and `eml` files because it spell _and_ grammar checks, which is what I want when I'm authoring those types of files. It is very good. 

I don't want to detract from what @kirawi said about "bloat"; I think that's an important discussion to have. Although I think it's also important to define what constitutes "bloat" before dismissing anything as too bloated. Perhaps `ltex-ls` is too bloated for some people, but like most software decisions, they are all tradeoffs, and what one loses in "bloat" may be sufficiently made up for in robustness and functionality for others. 

With that said, I don't think I'd want grammar checking in my code comments, as code comments tend to be held to a less rigorous grammatical standard, and reference code variables and things of that nature that may erroneously trigger spell and grammar checks.

**Reply by jerabaul29** [reply]
Nice, thanks for the advice @acaloiaro :) .

From my point of view I see 2 uses for spellcheck: I agree @acaloiaro that for "natural language" files (txt, md, a couple more), having advanced "spellcheck" including grammar is nice (thanks for the pointer, will give it a try :) ).

But I think that I also want "stupid simple" spellcheck on my code files. For example, just making sure that ```start_index``` is well spelled is quite useful: first, I do not like typos, second, I can quickly end up (especially depending on shadowing rules) with a ```strat_index``` somewhere for example, and this way failing to update ```start_index``` as I expected or this kind of things :) . I like to catch this with a spellchecker in all my files (ok, except assembly or other weird uses :) ).

**Reply by the-mikedavis** [reply]
Yep, the repo is private for now and I'm squatting the crate name :P. Spellbook can currently spell-check reasonably well (missing some features for compatibility with aspell) but suggesting corrections will take a while to work out. We may want to just start with showing spelling errors and work on suggestions in a follow-up effort. I'm a bit busy/distracted for most of November but I will make that repo public and push a minimal integration branch when I get the chance. In the meantime, I have seen people using and liking ltex-ls as mentioned above.

And yep, I'm imagining that we will want to add an indicator to `--health` to show whether you have a/any dictionary/ies installed.

**Reply by jerabaul29** [reply]
Sounds nice, I agree pushing a MVP even if it is a bit limited is better than nothing, and it makes it easier to get feedback / iterate as usual :) .

**Reply by akhilman** [reply]
I have managed to add CSpell as a spell checker using diagnostic-language-server. Both run by Deno runtime.
Here is my languages.conf:
```toml
# Languages

[[language]]
name = "git-commit"
language-servers = [ "dls" ]

[[language]]
name = "markdown"
language-servers = [ "marksman", "dls" ]

[[language]]
name = "rust"
language-servers = [ "rust-analyzer", "dls" ]

# Language servers

# Diagnostic language server.
# See https://github.com/iamcco/diagnostic-languageserver .

[language-server.dls]
command = "deno"
args = [ "run", "--allow-read", "--allow-env", "--allow-run", "--allow-sys", "npm:diagnostic-languageserver", "--stdio", "--log-level=2" ]

[language-server.dls.config.filetypes]
"*" = "cspell"

# CSpell https://cspell.org/docs/getting-started/ .

[language-server.dls.config.linters.cspell]
command = "deno"
args = [ "run", "--allow-read", "--allow-env", "npm:cspell", "stdin" ]
sourceName = "cspell"
rootPatterns = [
  ".cspell.json",
  "cspell.json",
  ".cSpell.json",
  "cSpell.json",
  "cspell.config.js",
  "cspell.config.cjs",
  "cspell.config.json",
  "cspell.config.yaml",
  "cspell.config.yml",
  "cspell.yaml",
  "cspell.yml",
]
debounce = 1000
formatLines = 1
formatPattern = [
  "^[^:]+:(\\d+):(\\d+) - (.*) \\((.*)\\)$",
  { line = 1, column = 2, security = 3, message = [3, " (", 4, ")"] },
]
securities = { "Unknown word" = "warning" }
```

**Reply by weakish** [reply]
@akhilman Does diagnostic-language-server support code actions (e.g. replace the word with suggestions, add word to dictionary, etc)?

**Reply by jerabaul29** [reply]
> Yep, the repo is private for now and I'm squatting the crate name :P. Spellbook can currently spell-check reasonably well (missing some features for compatibility with aspell) but suggesting corrections will take a while to work out.  [...]

@the-mikedavis just being curious: would you be ok to make the corresponding repo public, even if it is just in a very early form? I am trying to look a bit about the different options available and it would be very useful to see your work on this, even if it is preliminary :) (for example, I am still a bit unsure if I want to import a big LSP that offers suggestions, or if I am happy to start with something small that just highlights misspellings that I then 99% of the time can fix myself easily :) ).

**Reply by the-mikedavis** [reply]
I'm working on Spellbook actively again but I'm rewriting it based on [Nuspell](https://github.com/nuspell/nuspell) rather than [Spylls](https://github.com/zverok/spylls) for performance and correctness. The part about "Spellbook can currently spell-check reasonably well" is not the case at the moment but my plan is to open up the repo after the check algorithm is done. Without a full implementation of check though the project is pretty useless. There would be way too many false positives.

Suggestions and personal dictionaries could be done as follow-up. Plus the suggestion algorithm is simpler than check. My plan is to start small with a `:spell` typable command that segments the words in the current view, or maybe the whole buffer. But this is still a ways out so for now you're better off relying on a language server.

**Reply by gauteh** [reply]
Hey, thanks for this great project! Want to add my name to list of people anxiously awaiting this functionality (no pressure..). Again thanks for the great work!

**Reply by jerabaul29** [reply]
For now using https://github.com/blopker/codebook/issues/30#issuecomment-2708964041 , and it works well! Easy to install and set up.

**Comment by akhilman** [comment]
Nope, it does not. New words could be added to a custom dictionary file
in the project directory.

**Comment by David-Else** [comment]
Vale now has a language server, and tree sitter based spell checking for comments in programming languages like Rust!

https://github.com/errata-ai/vale/releases/tag/v3.5.0

https://github.com/errata-ai/vale-ls (needs vale installed)

I have not tried it in Helix yet, but it looks like it could be perfect.

**Comment by sbromberger** [comment]
I got vale working in Helix. It's really good.

One other one to try out is https://github.com/tekumara/typos-lsp - I was using this beforehand. It has really low false-positive rate but a higher false-negative rate due to the fact that it has a predefined list of common misspellings and only alerts on those. It's fairly comprehensive, though!

**Reply by mandarvaze** [reply]
@sbromberger 

> I got vale working in Helix. It's really good.

Do you mind sharing details/links/helix config ?

**Reply by sbromberger** [reply]
Sure. Here's the `languages.toml` entry:

```
[language-server.vale]
command = "vale-ls"
```

Then you just add it to whatever language config you want.

As far as installation goes, you'll need both vale and vale-lsp binaries on your path. I used a release of vale from the github releases page, but built vale-lsp from a git clone (using `cargo install .` from the cloned source directory). Then I created `~/.config/vale/.vale.ini` using the contents of the configuration generator (https://vale.sh/generator), and ran `vale sync` from the terminal. Everything just worked.

**Reply by David-Else** [reply]
I added it to the WIKI: https://github.com/helix-editor/helix/wiki/Language-Server-Configurations#vale-ls

**Reply by jsco2t** [reply]
Has anyone else had luck getting this to work?

I have both `vale` and `vale-ls` installed. Both are in my path.

I have these in my `languages.toml` file:

```toml
[language-server.vale]
command = "vale-ls"

[[language]]
name = "markdown"
language-servers = ["vale"]
text-width = 100
rulers = [100]
```

When I open a Markdown file - it doesn't appear that Vale is running as an LSP. I can misspell words, write incorrect grammer - nothing happens.

When I check the log file for Helix - this is the only message that I see:

```log
2024-09-04T11:35:23.779 helix_term::application [WARN] unhandled window/showMessage: ShowMessageParams { typ: Error, message: "missing field `Path` at line 4 column 1" }
```

If I ask Helix to restart the `lsp` using `:lsp-restart` the message in the editor UI is "Language server exited". The helix log is updated with:

```log
2024-09-04T11:45:59.548 helix_lsp [ERROR] client was already removed
2024-09-04T11:45:59.548 helix_lsp::transport [ERROR] vale err: <- StreamClosed
2024-09-04T11:45:59.668 helix_term::application [WARN] unhandled window/showMessage: ShowMessageParams { typ: Error, message: "missing field `Path` at line 4 column 1" }
```

Version info if:
```txt
❯ hx --version
helix 24.7 (079f5442)

❯ vale --version
vale version 3.7.0

❯ vale-ls --version
vale-ls 0.3.8
```

**Reply by jsco2t** [reply]
I suspect that maybe `vale-ls` needs parameters to work correctly? Though that's really just a guess on my part. I could have something else messed up here.

**Reply by Rot127** [reply]
You need to add a `.vale.ini` file. No idea why they don't just provide a default one.
```
cat ~/.config/vale/.vale.ini 
MinAlertLevel = suggestion

[*]
BasedOnStyles = Vale
```
Edit: Ok, just saw you fixed it below.

**Comment by sbromberger** [comment]
I'm using `vale-ls 0.3.7` and `vale version 3.6.0` and all I have in my languages.toml is 
```
[language-server.vale]
command = "vale-ls"
```

and things appear to be working.

**Reply by jsco2t** [reply]
I think I worked out my issue. To be clear - I have no idea if it's due to the versions of `vale/vale-ls` I have installed.

For whatever reason `vale` couldn't find the `.vale.ini` file if I put the file in: `~/.config/vale/.vale.ini`. I could only get things to work if I put the `.vale.ini` file in my _"home directory"_ ( `~/.vale.ini`).

I verified this by running `vale` directly. When I had my Vale config file in directory `~/.config/vale/.vale.ini` I could only get vale to run correctly if I explicitly specified the config file path:

```
vale README.md --output=JSON --config="$HOME/.config/vale/.vale.ini"
```

My fix (as mentioned above) was to create a `.vale.ini` file in my home directory. I didn't want vale to create it's artifacts in my home directory though - so I adjusted the styles path in it as follows:

```
❯ cat .vale.ini
StylesPath = .config/vale/styles

....file continues
```

**Comment by sbromberger** [comment]
I do not have a `~/.vale.ini`. I do have `~/.config/vale` which contains `.vale.ini` and `styles/` and `native/` subdirectories, both empty.

My `~/.config/vale/.vale.ini` contains
```
StylesPath = styles
MinAlertLevel = suggestion
Packages = Google, Readability

[*]
BasedOnStyles = Vale, Google, Readability
```

**Reply by jsco2t** [reply]
That's super interesting. Thanks for sharing @sbromberger. No idea why my environment works differently. 

```
¯\_(ツ)_/¯
```

Appreciate you following up!

**Comment by sbromberger** [comment]
Use the configuration tool (https://vale.sh/generator) to generate your `.vale.ini` file :)

**Reply by jsco2t** [reply]
I did use that :)

**Reply by jsco2t** [reply]
To be clear - I don't think the issue was the `.vale.ini` content. I *think* (for reasons I don't completely understand) in my environment Vale was not locating the `.vale.ini` when it was placed under `~/.config/vale/.vale.ini`

**Comment by joshbainbridge** [comment]
One other option that was recently shared on the Helix Matrix chat is [Harper](https://github.com/elijah-potter/harper/tree/master). This is the best LSP solution that I've tested so far. It's very fast and can be installed using cargo. Will work for both text based documents as well as comments. Another good option until we have @the-mikedavis's solution.

Configuration is in the wiki: https://github.com/helix-editor/helix/wiki/Language-Server-Configurations#harper-ls

**Reply by jsco2t** [reply]
Cool - I will have to check this out!

Thanks for passing along.

**Reply by masriomarm** [reply]
Have you made it work with git-commit? It works well with md files but can't work with git-commit.

**Reply by joshbainbridge** [reply]
Good point. I just did my first commit with it and I also didn't see it correct any spelling.

**Reply by sbromberger** [reply]
Sorry to ask an unrelated question here, but for those of you who have tried both Harper and Vale, could you post your thoughts comparing the two?

**Reply by sbromberger** [reply]
It seems to work with markdown, but doesn't work with .txt. I wonder why that is.

**Reply by jameynakama** [reply]
I can't get Harper to do anything in Python files. Green check mark in `hx --health` and my settings seem correct. But I just don't see it doing anything.

**Reply by sbromberger** [reply]
Are you passing `--stdio` in the arguments?

Specifically: https://github.com/elijah-potter/harper/discussions/150#discussioncomment-10594290

**Reply by jameynakama** [reply]
Yes. 

```
[[language]]
name = "python"
language-servers = ["pylsp", "ruff", "harper-ls"]
auto-format = true

[language-server.pylsp.config.pylsp]
plugins.pylsp_mypy.enabled = true
plugins.pylsp_mypy.live_mode = true

[language-server.ruff]
command = "ruff"
args = ["server"]

[language-server.harper-ls]
command = "harper-ls"
args = ["--stdio"]
```

**Comment by kirawi** [comment]
https://github.com/helix-editor/helix/issues/11660

**Reply by sbromberger** [reply]
Since we've been encouraged to take non-technical discussion of this PR here, I figured I'd share my thoughts:

I was initially very supportive of having a spellchecker integrated into Helix. However, I've since discovered several language servers that do everything I need - and more - across a wide variety of files.

One thing I didn't know I needed was grammar checking. This has turned out to be extremely useful and is now something I sort of want all over the place. It doesn't look like the tree-sitter/integrated effort has grammar checks, so it's now of limited utility. This might be by design, but at this point an integrated spell-check-only feature is something that I just wouldn't use and would possibly get in the way of my (working) setup. I would hope it's easy to disable and doesn't add too much bloat to the editor itself.

I don't want to be too discouraging, as I fully acknowledge my expectations / requirements have changed, and I realize this was a lot of work.

**Reply by jerabaul29** [reply]
Maybe it would make sense to have a bare bone, built in, ultra fast, native, simple spellchecker (i.e. the one listed in #11660 ) integrated into helix by default to make sure everybody using helix has basic typo correction support, but still encourage to use advanced checking through LSP for the users who have the time for a bit of setup? :)

For example, when just doing a bit of work fast on a server, it is nice that that barebone helix comes with a spellchecker, without the need to install plenty of additional LSP software :) .

**Reply by sbromberger** [reply]
@jerabaul29 - I'm not suggesting removing the functionality. I'd just like the ability to 1) turn it off, and 2) delete any wordfiles / dictionary files so that I'm not dealing with extra bloat. The language server approaches have their own dictionaries, after all.

**Reply by David-Else** [reply]
@sbromberger If you are using Harper, you can disable its spell-checking feature to get the best of both worlds: Helix for spelling and Harper for grammar.:

```
lspconfig.harper_ls.setup {
  settings = {
    ["harper-ls"] = {
      linters = {
        spell_check = false,
      }
    }
  },
}
```
I think you can do the same with Vale.

I would also keep using Vale for the writing style linting and grammar, but I imagine that the Helix spell checker will be better as the core devs are code Gods :)

**Reply by sbromberger** [reply]
@David-Else - that's a pretty creative solution, but if I'm gonna be running an extra LSP process, I might just as well go all-in.

I agree that Helix is an amazing bit of code, and I am grateful to the devs for making it as good as it is. I'm just sharing my opinion of the utility of the integrated spell-checker. I certainly don't mean to disparage the great work that's been done on the editor.

**Reply by masriomarm** [reply]
@David-Else any tips for making harper work with git-commit? It works normally with md files but not git-commit.

**Reply by sbromberger** [reply]
@masriomarm - sounds like the limitation is within harper. See https://github.com/elijah-potter/harper/issues/149 for what is likely the related issue.

**Reply by leo-unglaub** [reply]
I don't get why people are getting so bend out of shape about this feature. The history of Helix is one of the developers making good decisions and keeping stuff extremly usable and very well integrated. There is no reason to beleave that this is going to be any different. 

If you have doubts, just read the source code of https://github.com/helix-editor/spellbook, its very well written and extremly fast. So having a spell checker that works out of the box, amazing! 

Big thanks to @the-mikedavis for making this crate. Amazing work!

**Reply by sbromberger** [reply]
I don't see any evidence that people are getting bent out of shape. 

For my part, I am a bit concerned that this will introduce bloat (in the form of dictionaries) that will not be really useful to me given the alternatives, and wish to request the ability to disable the feature and remove any dictionaries so that I can continue to use this on small systems.

I certainly don't begrudge anyone from working on whatever features in Helix are important to them - that's the beauty of open source, and the Helix devs have indeed made good decisions in the past. But we were encouraged to take non-technical discussion of this feature to this thread, so people are sharing their opinions/wishes for this functionality.

I don't see anything wrong or unhealthy with that, and I'm glad that this feature will be of benefit to you and others.

**Reply by the-mikedavis** [reply]
Limiting the impact of builtin spellchecking seems like a fair ask to me and I was conscious of it while writing Spellbook. The only required dependency is hashbrown which we already depend on (and is a dep of Rust's stdlib) and the code itself is at-time-of-writing ~6k SLOC including tests, so the increase in compile time should be unnoticeable. I was envisioning the feature as off-by-default: either you would need to configure a locale plus install the dictionary or a simple config flag to enable a redistributed copy of en_US. I assume most people don't want their code spellchecked by default since there are lots of abbreviations used in code like "idx". I don't have strong opinions about including a default dictionary (almost certainly whatever is the most current en_US) in packages but if we did, en_US is a meager ~500Kb disk and wouldn't be loaded into memory unless the feature was enabled. Most likely we can take advantage of system-installed dictionaries like Kakoune does with its official aspell plugin.

At runtime builtin spellchecking should have no impact at all if disabled - we would skip loading the dictionaries, running tree-sitter queries, tokenizing and checking words.

**Reply by leo-unglaub** [reply]
@the-mikedavis what is your idea/plan for enabeling it? Something like this? Would you also put it into `languages.toml`? That way people could enable it automatically for certain types (git commit messages for example?) if they want to.

```
:set spellchecking true		# enables all found languages
:set spellchecking de_at	# enables German (Austria) language checking 
:set spellchecking en_us	# enables English (US) language checking
:set spellchecking false	# disables spell checking
```

Where would you prefer to have the language files stored? Lookup in the system directory (Firefox, Thunderbird, ... would share them) and then in `~/.dict/{en_us,de_at}`? Or would you prefer `<helix-dir>runtime/dict/*`?

**Reply by the-mikedavis** [reply]
How configuration works is worth some discussion (and I think would fit well as comments in #11660). You would need to set a locale for a document or a language, or a global locale, also see https://github.com/helix-editor/helix/issues/8853#issuecomment-1887685024. You might want multiple locales - one for commit messages and one for code for example. It might make sense to also have a config flag that flat enables/disables the feature.

For dictionary files I think there is something "standard" - i.e. where LibreOffice will store them. Firefox maintains its own en_US but we could try to find that too. Kakoune's aspell integration finds installed ones and I believe ZSpell has some code to find them as well: https://github.com/pluots/zspell. But this part needs some further investigation.

**Reply by justinlovinger** [reply]
> I imagine that the Helix spell checker will be better as the core devs are code Gods :)

If the devs want to make a better spellchecker, they can make a language-server everyone can use, not just Helix. It could ship with Helix and languages could have it as a default language-server. Being a language-server will not stop it from using tree-sitter to spellcheck in code.

**Reply by justinlovinger** [reply]
> I don't get why people are getting so bend out of shape about this feature.

As one of the dissenting voices, I am going to echo what @sbromberger said, I think "getting bent out of shape" is overly strong language. I disagree with the decision and am discussing it in the hopes of coming to a better understanding for everyone, myself included. I will continue to use Helix regardless of the outcome.

**Reply by the-mikedavis** [reply]
Splitting out a spellchecker into a separate project that communicates over LSP is a viable alternative but it would mean duplicating a lot code and also needlessly duplicating memory and work at runtime. Plus it means overhead of configuring and installing a language server.

**Reply by sbromberger** [reply]
@the-mikedavis not to mention, it's already been done by several other packages.

**Reply by justinlovinger** [reply]
> it would mean duplicating a lot code

I imagine organizing the tree-sitter code into a common library could eliminate most of the code-duplication. It could be a good refactor anyway.

> also needlessly duplicating memory and work at runtime

How much runtime overhead would we be looking at in practice? It is my understanding tree-sitter is very fast and relatively lightweight.

On a related note, once we get plugin-support, spellchecking could be a good candidate for a plugin. It could ship with Helix and still be enabled by default.

**Reply by kirawi** [reply]
> I imagine organizing the tree-sitter code into a common library could eliminate most of the code-duplication. It could be a good refactor anyway.

This is already happening by the way

**Comment by justinlovinger** [comment]
<https://github.com/helix-editor/helix/issues/11660#issuecomment-2339311940>

> I spent a non-trivial bit of time setting up LSP's (in my case Vale) just so that I could get spell-checking in Markdown docs.

Your mistake may have been trying Vale first. Vale is significantly harder to set-up than Harper. Better documentation for how to set-up a spellchecking LSP could alleviate this issue.

**Reply by jsco2t** [reply]
I accept that - and as it happens now that I have used Harper I agree that it's a solid solution for Grammar + Spelling. That being said. I would still vastly prefer a native solution for spell checking. A native solution should open up more opportunities in terms of where spell checking can happen while also allowing the spell-checking solution to be highly optimized.

Just my 2 cents. I totally get others have a different perspective on this.

**Reply by jsco2t** [reply]
Also - I really appreciate the follow-up. I missed mentioning that above.

**Reply by justinlovinger** [reply]
> A native solution should open up more opportunities in terms of where spell checking can happen while also allowing the spell-checking solution to be highly optimized.

LSP-based spellcheckers can also use tree-sitter for spellchecking in code, and many do. Helix already plans to splits its tree-sitter code into a library. At that point, LSP-based spellcheckers could use the exact same code-traversal Helix would use for spellchecking.

The only actual downside to using a LSP-based spellchecker is performing tree-sitter calculations twice. However, given tree-sitter is incremental and very efficient, it is not clear how much of a downside that actually is in practice.

**Reply by the-mikedavis** [reply]
The time and memory varies with the different parsers/scanners of each language, as well as disk space of a parser - some are fairly large. In the common case it's pretty fast but there are degenerate cases that can take hundreds of milliseconds to parse. A language server would also mean duplicating the document in memory between Helix and the server. It's definitely possible but needlessly wasteful and - again - a worse experience as you would need to configure and install an extra server. I don't see any advantage in scoping this as a plugin: it would mean either shelling out to Hunspell/Nuspell/Aspell (this is what Kakoune does) or writing a new spellchecking library in Scheme.

My take is that basic spellchecking fits as a core feature. It's a subjective judgement and disagreeing is fine, though this thread seems to me to be falling into whataboutism and hand-wringing. It's an additive feature with a reasonably small footprint that can benefit anyone who doesn't already have a prose checking setup, is potentially capable of spellchecking within code (not comments), and doesn't threaten existing prose-checking setups.

**Comment by justinlovinger** [comment]
<https://github.com/helix-editor/helix/issues/11660#issuecomment-2340191133>

> With helix we are in a position where we do have a universal parser for every language and that means we can enable a much more reliable spellchecker.

Harper also uses tree-sitter for spellchecking in code, just like Helix would.

**Reply by David-Else** [reply]
Vale also uses tree-sitter: https://github.com/errata-ai/vale/releases/tag/v3.5.0

> This release includes improved, [tree-sitter](https://tree-sitter.github.io/tree-sitter/) powered support for linting comments in Go, Rust, Python, Ruby, C/C++, JavaScript, TypeScript, YAML, and CSS files.

**Reply by the-mikedavis** [reply]
Both support a relatively limited set of tree-sitter parsers compared to our language support and only use tree-sitter to find comment nodes. We should also aim to find typos in code like struct fields for example.

**Comment by hholst80** [comment]
The point of Helix is that batteries ARE INCLUDED.  Spell check should be a feature that comes out of the box.  It certainly does that from its big brother Vim.

If it is not included users might as well use Vim + vendor meta package like AstroNvim.

**Reply by kirawi** [reply]
https://github.com/helix-editor/helix/issues/11660

**Comment by jerabaul29** [comment]
I am now using codebook https://github.com/blopker/codebook , which takes spellbook https://github.com/helix-editor/spellbook and turns it into a LSP, while integrating with treesitter, logics with camelcase etc. Works like a charm, I can recommend. See the installation instructions at https://github.com/blopker/codebook/issues/30#issuecomment-2708964041 .
